src/Planners/src/UnicyclePlanner.cpp

```diff
@@ -0,0 +1,600 @@
+{
+    constexpr auto logPrefix = "[UnicyclePlanner::advance]";
+
+    if (!m_pImpl->planner)
+    {
+        log()->error("{} The Unicycle planner has never been initialized.", logPrefix);
+        return false;
+    }
+
+    m_pImpl->output = std::nullopt;
+
+    // Initialize the left FootPrint
+    m_pImpl->left = std::make_shared<FootPrint>();
+    m_pImpl->left->setFootName(m_pImpl->names.left);
+
+    // Initialize the right FootPrint
+    m_pImpl->right = std::make_shared<FootPrint>();
+    m_pImpl->right->setFootName(m_pImpl->names.right);
+
+    // Compute the FootPrints
+    if (!m_pImpl->planner->computeNewSteps(m_pImpl->left,
+                                           m_pImpl->right,
+                                           m_pImpl->horizon.t0,
+                                           m_pImpl->horizon.tf))
+    {
+        m_pImpl->left = nullptr;
+        m_pImpl->right = nullptr;
+        log()->error("{} Failed to compute new steps.", logPrefix);
+        return false;
+    }
+
+    // Create and configure the generator
+    auto generator = UnicycleGenerator();
+    generator.setSwitchOverSwingRatio(m_pImpl->gait.stancePhaseRatio);
+    double lastStepSwitchTime = 0.8;
+    generator.setTerminalHalfSwitchTime(lastStepSwitchTime);
+    generator.setPauseConditions(m_pImpl->duration.max, m_pImpl->duration.nominal);
+
+    // Compute the contact states using the generator
+    if (!generator.generateFromFootPrints(m_pImpl->left,
+                                          m_pImpl->right,
+                                          m_pImpl->horizon.t0,
+                                          m_pImpl->dt.planner))
+    {
+        log()->error("{} Failed to generate from footprints.", logPrefix);
+        return false;
+    }
+
+    // Get the contact states over the horizon
+    std::vector<bool> leftStandingPeriod;
+    std::vector<bool> rightStandingPeriod;
+    generator.getFeetStandingPeriods(leftStandingPeriod, rightStandingPeriod);
+
+    // Lambda to convert the vector with contact states to a vector of PlannedContact.
+    // It just processes timings, we fill the transform later.
+    auto processStandingPeriod
+        = [](const std::vector<bool>& isFootInContactVector,
+             const double dt,
+             const std::string& contactName) -> std::vector<Contacts::PlannedContact> {
+        std::vector<Contacts::PlannedContact> contacts;
+
+        // Loop over all the horizon
+        for (size_t i = 1; i < isFootInContactVector.size(); ++i)
+        {
+            const bool thisState = isFootInContactVector[i];
+            const bool lastState = isFootInContactVector[i - 1];
+            const bool aboutToLand = lastState == 0 && thisState == 1;
+
+            // Create a new contact
+            if (i == 1 || aboutToLand)
+            {
+                auto contact = Contacts::PlannedContact();
+                contact.name = contactName;
+                contact.activationTime = dt * i;
```

Comment:
I would use the activation time provided by the planner instead of computing again. This is because the activation time of a step can be smaller than the init time specified to the planner.

Reply:
Yes something could be off here, I'm not sure what's the intended behavior when `m_pImpl.horizon.t0 != 0`. Do you think that the following would be better?

```diff
-               contact.activationTime = dt * i;
+               contact.activationTime = m_pImpl.horizon.t0 + dt * i;
```

Reply:
Why can't we just use the computed ``impactTime``?

Reply:
Yes, this seems actually a good idea. This for loop would only process the creation of a new step without the complexity of also computing the duration.